Feat 52 search by genre

xxxserxxx opened the pull request:
# This is forked off issue-76-clarify-search-results!!!

It's branched here because, otherwise, the merge for both will be nasty. It's a **separate** branch, because it adds a feature.

This implements #52. 

- It works better with the feature that focuses the columns rather than the search field.
- It's _in_ the search page; this is largely to prevent the proliferation of pages
- Pressing 'g' in any column toggles between genre search, and a general `search3` search
- Genre searches _only_ return songs (that's an API thing). This means the artist and album fields are unused, which leads to....
- in genre-search mode, the album field is renamed to "genres" and is filled with a list of genres the server knows, pulled by `/getGenres`
- Selecting a genre from the middle column is the same as searching for that genre, and fills the songs column with songs with that genre.
- Genre searches are case sensitive, on both gonic and Navidrome.
- An entire genre can be added to the queue with `a`

I try to keep commits focused, but this commit changes a couple of other (unrelated) things:

- I added Year to the song info metadata. I originally wanted to include the genre, but most of the API calls returning song info do _not_ include the genre with the song; when included, it's in the album info. A work-around would have been far more invasive, so I've put it off, but I kept the Year change.
- The Queue page now show the size of the queue in the column title. I originally did this to test that queue population was adding the right number of things, and because
- I started expanding column counts in the search page, and liked it. I think ultimately I'd like to add it to **all** lists.

You _should_ be able to merge "clarify" and then merge this branch; or if you decide to merge both, just merge this branch and you'll get both.

xxxserxxx committed:
feature: #76, clarify search results

xxxserxxx committed:
add: search now pro-actively searches until no more results are returned. It still queries in batches of 20, and updates the list(s) as results are available.

xxxserxxx committed:
fix: clarifies search help text

xxxserxxx committed:
feat: add entire genre; show item counts in queue and genre columns

xxxserxxx commented:
@spezifisch There are conflicts between my PRs, because I keep forking them off main and they often overlap files. When it's too obvious to me that merges are going to be hard this way, I'll branch off another of my unmerged branches, but this has its own problems.

I have a suggestion: when you're next ready, tell me which of the PRs you would like to merge. I'll create a merge branch off main, merge and resolve the conflicts, and you can take a look at it, test run it, and pull them in at once without having to deal with the conflicts.

spezifisch commented:
> I have a suggestion: when you're next ready, tell me which of the PRs you would like to merge. I'll create a merge branch off main, merge and resolve the conflicts, and you can take a look at it, test run it, and pull them in at once without having to deal with the conflicts.

Let's try that. I've just been through PRs and want to weed out the performance bugs on my setup next.
This PR here is the one I would merge next.

xxxserxxx committed:
feat: #52, search by genre

xxxserxxx committed:
fix: clear search columns when genre search mode changes
fix: make sure _something_ other than the search field gets focused on Enter

xxxserxxx commented:
> Let's try that. I've just been through PRs and want to weed out the performance bugs on my setup next. This PR here is the one I would merge next.

Yeah, that was an ugly one.

I rebased the branch on the current main; it passes `make all` and I cleaned up a couple of column content things in 5cac4ac. Then I did a test merge back onto main with no issues.

I have yet to remember to sign a commit :-/

xxxserxxx commented:
@spezifisch how does this new patch look?

mahmed2000 commented:
Initially mentioned this in the Album-blanking thread. Putting it here for clarity.

Toggling the genre-mode mid-search results in some odd behavior. Video demonstrating this below. Only control inputs are "enter" -> "g" on the first search and on "Soundtrack" on the second.

[Search interruption oddities](https://github.com/user-attachments/assets/1e0fb400-6635-4228-a40c-6be72bfdb559)

For reference, there is a way to simulate latency on a local connection by adding a delay to an interface with `tc qdisc` on linux, even for loopback. Might be useful for debugging but might also be useful for tests in general.
